### PR TITLE
object-oriented-js (frogger) - done

### PR DESCRIPTION
**Pull request by Karasb888**
object-oriented-js (frogger)

[Demo](https://karasb888.github.io/frogger-game/) |
[Code base](https://github.com/Karasb888/frogger-game/blob/master/js/app.js)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Comment by OleksiyRudenko** [comment]
@Karasb888 Looks pretty good at the first glance.

Please, self check yourself against the list below:

 * [ ] it is OK to employ ES6 features like `const`, `let` etc.
 * [ ] OO is implemented using JS prototype chain object model (**not** ES6 classes syntax)
 * Requirements re **Constants**:
   * [ ] all numbers like block dimensions, initial locations are defined as constants
   * [ ] there are core contants and derived constants
         (e.g. `const FIELD_WIDTH = BLOCK_WIDTH * BLOCKS_NUMBER;`)
   * [ ] arrays of constants are also constants 
         (e.g. `const INITIAL_POSITIONS = [1,2,3,4].map(rowNumber => rowNumber * BLOCK_HEIGHT);`)
   * [ ] const objects help organizing and structure const data even better
         (e.g. `const PLAYER_CONF = { initialPosition: {x: 1, y: 5}, sprite: '...', ...etc... };`
 * Requirements re **OOP**:
   * [ ] properties common for some classes are generalized into a base class
         (e.g. there is `Character` base class, which is extended by `Enemy` and `Player` classes)
   * [ ] class extension is implemented using `Subclass.prototype = Object.create(Superclass.prototype)`,
         not `Subclass.prototype = new Superclass(params);`;
         [Useful resource](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)
   * [ ] classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
         (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
         Hint: pass `Player` instance as an argument to every enemy
   * [ ] Separation of Concerns principle is followed
         (e.g. `update` method does only rendering and 
         doesn't **contain** any inline code to check e.g. collisions)
   * [ ] the code is very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
 * Most common mistakes
   * [ ] Make sure `target = condition ? valueWhenConditionTrue : valueWhenConditionFalse` is used instead of
         `condition ? target = valueWhenConditionTrue : target = valueWhenConditionFalse`;
         [Conditional (ternary) operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator)
         
Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

**Comment by Karasb888** [comment]
I hurried with OOP, that was rly bad idea to try end this task (and another one) in one night=). After second look I see that I have global mistakes and need to improve my knowledge of OOP=)

**Comment by Karasb888** [comment]
Will fix it as soon as I can=)

**Comment by OleksiyRudenko** [comment]
@Karasb888 This task code is really good. Many things from the list seem to be implemented.
Just go through the code and check against the list to make sure everything is taken into the account.

**Comment by Karasb888** [comment]
So, I do my best =)

File: submissions/Karasb888/object-oriented-js/app.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    table.innerHTML = 'YOU LOSE';
+    alert('YOU LOSE');
+  }else{
+  table.innerHTML = 'Lives: ' + lives + ' | Wins: ' + wins;
+  }
+
+};
+
+document.addEventListener('DOMContentLoaded', function(){
+
```
Comment (OleksiyRudenko) [inline submissions/Karasb888/object-oriented-js/app.js]: Empty lines right after function signature (and right before closing `}` at function end) do not add to code readability. Please, remove here and across entire code base.

File: submissions/Karasb888/object-oriented-js/app.js [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+  const table = document.getElementById('table');
+  setScore();
+
+});
+
+//MAIN constans
+var GAME_PARAMS = {
+  canvas_width: 505,
```
Comment (OleksiyRudenko) [inline submissions/Karasb888/object-oriented-js/app.js]: In JS snake case (`abc_def`) is used when entity name is all caps. Please, apply camel case here.
All caps prefix (e..g. `GAME_PARAMS` in `GAME_PARAMS.canvasWidth`) clearly tells we deal with a constant.

File: submissions/Karasb888/object-oriented-js/app.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  canvas_height: 606,
+  row_height: 83,
+  coll_width: 101,
+  coll_numb: 6,
```
Comment (OleksiyRudenko) [inline submissions/Karasb888/object-oriented-js/app.js]: Pretty ambiguous shortening. Do not shorten words when it save 2 or 3 characters. 
Does `coll` stand for `collocation`, `collection` or `column`?
We write code for other people, firsthand. Some of them will be seeing for the first time.

File: submissions/Karasb888/object-oriented-js/app.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+Enemy.prototype = Object.create(Character.prototype);
+
+Enemy.prototype.start = function() {
+  this.x = -GAME_PARAMS.coll_width;
+  this.speed = getRandomArbitrary(1, 6);
+};
+
+Enemy.prototype.collision = function() {
+  //On collision with player
+  if(this.x  >= player.x - 70 && this.x - 70 <= player.x && this.row === player.row ){
```
Comment (OleksiyRudenko) [inline submissions/Karasb888/object-oriented-js/app.js]: `70` is a magic number. Please, define it as a constant.
Please, find other similar magic values across the code base and define those as constants.

File: submissions/Karasb888/object-oriented-js/app.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+      if(this.x + GAME_PARAMS.coll_width < GAME_PARAMS.canvas_width){
+        this.x += GAME_PARAMS.coll_width;
+      }
+    break;
+
+    case 'up':
+        this.y -= GAME_PARAMS.row_height;
+        this.row++;
+    break;
+
```
Comment (OleksiyRudenko) [inline submissions/Karasb888/object-oriented-js/app.js]: When having `case` fragments properly indented no need to have empty lines anywhere inside `switch`.
```
case 'up':
  this.y -= GAME_PARAMS.row_height;
  this.row++;
  break;
```

Please, install linters and code prettifying plugins to your IDE. They do all dirty job for you formatting code as per conventions.

File: submissions/Karasb888/object-oriented-js/app.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+};
+
+var player = new Player();
+
+var allEnemies = [];
+for (var i = 1; i <= 3; i ++){
+  allEnemies.push( new Enemy(i, player) );
+}
+// This listens for key presses and sends the keys to your
```
Comment (OleksiyRudenko) [inline submissions/Karasb888/object-oriented-js/app.js]: `Array` methods are powerful.
```
var allEnemies = [1, 2, 3].map(row => new Enemy(row, player));
```
You are not required to fix this. Keep the technique in mind.

File: submissions/Karasb888/object-oriented-js/app.js (hunk 4 above)
Comment (OleksiyRudenko) [reply submissions/Karasb888/object-oriented-js/app.js]: Also `player` here is a global variable.
Please, fix.

File: submissions/Karasb888/object-oriented-js/app.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+};
+
+Enemy.prototype = Object.create(Character.prototype);
+
+Enemy.prototype.start = function() {
+  this.x = -GAME_PARAMS.coll_width;
+  this.speed = getRandomArbitrary(1, 6);
+};
+
+Enemy.prototype.collision = function() {
```
Comment (OleksiyRudenko) [inline submissions/Karasb888/object-oriented-js/app.js]: By convention functions are actions and their names should start with a verb.
`collision` is a noun. Please, refactor.

**Pull request merged**